src/hexadecimal.jl

```diff
@@ -0,0 +1,160 @@
+# All bytes that are not hex digits are 0x0000FFFF
+const _HEX_LUT = UInt32[
+    b in UInt8('0'):UInt8('9') ? UInt32(b - UInt8('0')) :
+        b in UInt8('a'):UInt8('f') ? UInt32(b - UInt8('a') + 0x0a) :
+        b in UInt8('A'):UInt8('F') ? UInt32(b - UInt8('A') + 0x0a) :
+        0xFFFF
+    for b
+    in UInt8(0):UInt8(255)
+]
+
+### SHA1 ###
+
+struct SHA1
+    h::NTuple{5,UInt32}
+end
+
+supportedtype(::Type{SHA1}) = true
+
+_set_tup5_at(t::NTuple{5,UInt32}, i, x::UInt32) = ntuple(j -> j == i ? x : t[j], 5)
+
+function typeparser(::Type{SHA1}, source, pos, len, b, code, pl, options)
+    if len - pos + 1 < 40
+        @inbounds while pos <= len && _HEX_LUT[peekbyte(source, pos) + 0x01] != 0xFFFF
+            pos += 1
+            incr!(source)
+        end
+        eof(source, pos, len) && (code |= EOF)
+        return (pos, code | INVALID, PosLen(pl.pos, pos - pl.pos), SHA1((0,0,0,0,0)))
+    end
+
+    check = UInt32(0)
+    out = ntuple(j -> UInt32(0), 5)
+    @inbounds for i in 1:5
+        h = UInt32(0)
+        for _ in 0:7 # Maybe unroll?
+            b = peekbyte(source, pos)
+            c = _HEX_LUT[b + 0x01]
+            check |= c
+            h = ((h << 4) | c)
+            pos += 1
+            incr!(source)
+        end
+        # Only check validity every 8 bytes to hopefully enable some vectorization
+        if check == 0xFFFF
+            pos -= 8 # Backtrack to the start of the invalid hex
+            fastseek!(source, pos)
+            while _HEX_LUT[peekbyte(source, pos) + 0x01] != 0xFFFF
+                pos += 1
+                incr!(source)
+            end
+            return pos, code | INVALID, PosLen(pl.pos, pos - pl.pos), SHA1(out)
+        end
+        out = _set_tup5_at(out, i, h)
+    end
+    eof(source, pos, len) && (code |= EOF)
+    return (pos, code | OK, PosLen(pl.pos, pos - pl.pos), SHA1(out))
+end
+
+### UUID ###
+
+supportedtype(::Type{UUID}) = true
+
+function typeparser(::Type{UUID}, source, pos, len, b, code, pl, options)
+    check = UInt32(0)
+    hi = UInt64(0)
+    lo = UInt64(0)
+
+    @inbounds begin
```

Comment:
Can we please indent under this `begin` 

without it I found it hard to find where the `inbounds` ends (everytime i saw an `end` i had to check what it was for)